### PR TITLE
Methods to create an empty `TaskStatusDB`

### LINKED ISSUE #1: TaskStatusDB: Set up an empty database
Empty database includes two tables:

* `tasks` with columns:
  * `taskid: str`
  * `status: int`
  * `last_modified: datetime`
  * `tries: int`
* `dependencies` with columns:
  * `to: str` (FK on `tasks.taskid`)
  * `from: str` (FK on `tasks.taskid`)

Main user-facing to implement:

* `__init__(self, engine: sqla.Engine)`
* `from_filename(cls, filename: os.PathLike)`

### PR DESCRIPTION
**Pull request by dwhswenson**
Resolves #1 

Also includes the general skeleton of things to come in `TaskStatusDB`

### CODE REVIEW

**Comment by dwhswenson** [comment]
This is ready for review... the remaining missing coverage are things that I'll push out to beyond the scope of this PR.

File: exorcist/taskdb.py [hunk 1]
```diff
@@ -1,40 +1,162 @@
...
+        self.engine = engine
 
+    def __repr__(self):
+        return f"TaskStatusDB(engine={self.engine})"
 
-def load_task_db(engine):
-    ...
+    @property
+    def tasks_table(self):
+        return self.metadata.tables['tasks']
```
Comment (richardjgowers) [inline exorcist/taskdb.py:48]: what's the return type of this?

File: exorcist/taskdb.py [hunk 2]
```diff
@@ -1,40 +1,162 @@
...
+       fixed, if possible, and exit otherwise.
+    2. So that, in the event that the status that needs to be set on early
+       exit (e.g., due to a sigterm send by the queuing system) is the same
+       as the current status, it can be caught without error and exit
+       normally.
     """
-    Create an empty task database with the provided engine.
+
+
+class TaskStatusDB:
```
Comment (richardjgowers) [inline exorcist/taskdb.py:28]: can we make this the sqla version of an abc?

**Review by richardjgowers — CHANGES_REQUESTED** [review]
I don't want sqla being necessary for the API

**Comment by dwhswenson** [comment]
> I don't want sqla being necessary for the API

How are you planning for us to easily scale to use something other than SQLite?

File: exorcist/taskdb.py [hunk 3]
```diff
@@ -1,40 +1,162 @@
...
+        Raises
+        ------
+        NoStatusChange :
+            If the status does not actually change (e.g., the code is trying
+            to change it to a status it already has), this is raised. This
+            allows callers to catch and handle this case.
+        """
         ...
 
+    def mark_task_completed(self, completed_taskid: str):
```
Comment (richardjgowers) [inline exorcist/taskdb.py:157]: bool of success

File: exorcist/taskdb.py [hunk 4]
```diff
@@ -1,40 +1,162 @@
...
+        """
+        Update the database (including the DAG info) to show that the task
+        has been completed.
+        """
+        ...
```
Comment (richardjgowers) [inline exorcist/taskdb.py:162]: add `checkout_next_task() -> str`

File: exorcist/taskdb.py [hunk 5]
```diff
@@ -1,40 +1,162 @@
...
+
+        Parameters
+        ----------
+        task_network: nx.Digraph
+            A network with :class:`.Task`\ s as nodes.
+        """
         ...
 
-    def _load_tasks(self):
+    def update_task_status(
```
Comment (richardjgowers) [inline exorcist/taskdb.py:132]: possibly not required yet

File: exorcist/taskdb.py [hunk 6]
```diff
@@ -1,40 +1,162 @@
...
+        deps_table = sqla.Table(
+            "dependencies",
+            metadata,
+            sqla.Column("from", sqla.String, sqla.ForeignKey("tasks.taskid")),
+            sqla.Column("to", sqla.String, sqla.ForeignKey("tasks.taskid")),
+        )
+        # TODO: create indices that may be needed
+        metadata.create_all(bind=engine)
+
+    def add_task(self, task: Task, requirements: Iterable[Task]):
```
Comment (richardjgowers) [inline exorcist/taskdb.py:109]: task should be str

**Comment by dwhswenson** [comment]
@richardjgowers : To enable retries we need one of the following in the API around `check_out_task` / `mark_task_completed`:

1. `check_out_task` returns more than the task ID; at a minimum returns number of tries so far, but it would make more sense to return something that could be extended with other information in the future. Combine this with `mark_task_completed` taking a `TaskStatus` instead of a `bool` for success/fail. Logic of "have we hit max retries" is on the worker.
2. `mark_task_completed` takes in a `max_retries` parameter. This will require 2 calls to the DB; one to read the current number of retries, one to update. Logic on "have we hit max retries" is on the TaskStatusDB.
3. Another option: `max_retries` is a `@property` on the DB (must be given somehow at initialization). Like 2, this will require 2 DB calls. Retry counting on the DB.
4. An option to avoid the extra DB call (messier, IMO, once you get into the details): `check_out_task` gives a context manager, and `mark_task_completed` is just its cleanup phase. This sounded good, until I realized that the only ways I could imagine to communicate success/fail status are (a) yield a mutable object (instead of just the task ID str) and mutate that on the worker for success/failure; (b) raise a custom exception for failure results, and catch that in the context manager. Neither seemed very nice.
5. Another option to avoid the extra DB call: store the number of retries for this task as mutable state in the `TaskStatusDB`. This makes some assumptions, like that a given `TaskStatusDB` instance will only have one task in-flight at a time, but those are probably reasonable assumptions. However, I dislike mutable state as instance attributes, and this breaks some independence of the methods.

**Comment by richardjgowers** [comment]
@dwhswenson I think I'm leaning towards 5 if I've understood this.  I think for the retries we're talking about, I don't think a Worker needs to be aware that it is doing a retry of something.  Once we have Worker tracking and there's attempts to recover things that went dark that is a different case where the Worker does require further information.

I think for retries to work we're going to have to have some table of taskid:retry count.

I don't think this alters the signature of checkout/mark for a Worker since the logic for handling retries is all DB-side.

re: counting calls. It's ok if there's multiple calls to the DB if these calls are all happening in the body of the DB methods, there just needs to be a single call from Worker to do transactions so that Worker can die at any point and not leave the DB halfway.

**Comment by dwhswenson** [comment]
Add `max_retries` as column for each task

**Comment by dwhswenson** [comment]
@richardjgowers this is ready for another look. I also put in a `mark_aborted_incomplete` -- it feels to me like failures because something went wrong with the execution system (wall time failures, most likely) are qualitatively different from something going wrong during the task. But I'm not sure how much we want to distinguish those, except maybe by adding logging/output (2 retry counters seems heavy to me).

File: exorcist/models.py [hunk 7]
```diff
@@ -1,23 +1,31 @@
...
+TaskDetails = TypeVar("TaskDetails")
 
 class TaskStatus(Enum):
+    """
+    Status of a given task.
+    """
     BLOCKED = 0
     AVAILABLE = 1
     IN_PROGRESS = 2
     RESULTS_READY = 3
```
Comment (richardjgowers) [inline exorcist/models.py:16]: this status isn't planned any more right?

File: exorcist/models.py [hunk 8]
```diff
@@ -1,23 +1,31 @@
...
-class Task:
-    taskid: str
-    taskfile: str
-    last_modified: datetime
-    n_retries: int
+# TODO: it isn't entirely clear to me that this is needed, or that this is
+# the right way to do it. but I wanted to capture the way to handle typing
+# of something like this
+@dataclasses.dataclass
+class Task(Generic[TaskDetails]):
```
Comment (richardjgowers) [inline exorcist/models.py:26]: this class is going right?

File: exorcist/taskdb.py [hunk 9]
```diff
@@ -1,40 +1,226 @@
...
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def add_task_network(self, task_network: nx.DiGraph):
+        """Add a network of tasks to the database.
+
+        Parameters
+        ----------
+        task_network: nx.Digraph
+            A network with taskid strings as nodes.
```
Comment (richardjgowers) [inline exorcist/taskdb.py:51]: if we're doing this it's worth documenting which way round the directional edges work.  Do nodes point to children or parents?

File: exorcist/taskdb.py [hunk 10]
```diff
@@ -1,40 +1,226 @@
...
+        -------
+        str :
+            The taskid of the task to run
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def mark_task_aborted_incomplete(self, taskid: str):
+        """
+        Update the database when a task fails to complete.
```
Comment (richardjgowers) [inline exorcist/taskdb.py:73]: it's not clear how this is different from completed with success=False.  Does this not increment the number of failures?  It just reverts the task to its status before it was checked out?

**Review by richardjgowers — CHANGES_REQUESTED** [review]

File: exorcist/taskdb.py (hunk 10 above)
Comment (dwhswenson) [reply exorcist/taskdb.py:73]: See https://github.com/OpenFreeEnergy/exorcist/pull/9#issuecomment-1604727304

File: exorcist/taskdb.py (hunk 9 above)
Comment (dwhswenson) [reply exorcist/taskdb.py:51]: Part of #14. This is just supposed to be a rough skeleton here; let's worry about this when we implement these things.

File: exorcist/models.py (hunk 7 above)
Comment (dwhswenson) [reply exorcist/models.py]: removed in [240b325](https://github.com/OpenFreeEnergy/exorcist/pull/9/commits/240b3255297a57ba79900a462789673a14b58595)

File: exorcist/models.py (hunk 8 above)
Comment (dwhswenson) [reply exorcist/models.py]: removed in [240b325](https://github.com/OpenFreeEnergy/exorcist/pull/9/commits/240b3255297a57ba79900a462789673a14b58595)

**Comment by dwhswenson** [comment]
@richardjgowers anything else blocking here?

File: exorcist/models.py [hunk 11]
```diff
@@ -1,23 +1,19 @@
 from enum import Enum
-from dataclasses import dataclass
-from datetime import datetime
+from typing import TypeVar, Generic
+import dataclasses
+
+# generics: the actual types here depend on the client library
+Result = TypeVar("Result")
+TaskDetails = TypeVar("TaskDetails")
```
Comment (richardjgowers) [inline exorcist/models.py:7]: these are also both not required right?

File: exorcist/taskdb.py (hunk 10 above)
Comment (richardjgowers) [reply exorcist/taskdb.py:73]: So if that's the intent put it in the docstring, else it'll never happen.  If it's not thought out yet, don't include it yet.  We're going for minimal.

**Review by richardjgowers — APPROVED** [review]

**Pull request merged**